examples/se2_localization_liekf.cpp

```diff
@@ -0,0 +1,282 @@
+    U        = (u_sigmas * u_sigmas).matrix().asDiagonal();
+
+    // Declare the Jacobians of the motion wrt robot and control
+    manif::SE2d::Jacobian J_x, J_u, AdX, AdXinv;
+
+    // Define the gps measurements in R^2
+    Vector2d    y, y_noise;
+    Array2d     y_sigmas;
+    Matrix2d    R;
+
+    y_sigmas << 0.1, 0.1;
+    R        = (y_sigmas * y_sigmas).matrix().asDiagonal();
+
+    // Declare the Jacobian of the measurements wrt the robot pose
+    Matrix<double, 2, 3>    H;      // H = J_e_x
+
+    // Declare some temporaries
+    Vector2d                e, z;   // expectation, innovation
+    Matrix2d                E, Z;   // covariances of the above
+    Matrix<double, 3, 2>    K;      // Kalman gain
+    manif::SE2Tangentd      dx;     // optimal update step, or error-state
+
+    //
+    //
+    // CONFIGURATION DONE
+
+
+
+    // DEBUG
+    cout << std::fixed   << std::setprecision(3) << std::showpos << endl;
+    cout << "X STATE     :    X      Y    THETA" << endl;
+    cout << "----------------------------------" << endl;
+    cout << "X initial   : " << X_simulation.log().coeffs().transpose() << endl;
+    cout << "----------------------------------" << endl;
+    // END DEBUG
+
+
+
+
+    // START TEMPORAL LOOP
+    //
+    //
+
+    // Make 10 steps. Measure one GPS position each time.
+    for (int t = 0; t < 10; t++)
+    {
+        //// I. Simulation ###############################################################################
+
+        /// simulate noise
+        u_noise = u_sigmas * Array3d::Random();             // control noise
+        u_noisy = u_nom + u_noise;                          // noisy control
+
+        u_simu   = u_nom;
+        u_est    = u_noisy;
+        u_unfilt = u_noisy;
+
+        /// first we move - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+        X_simulation = X_simulation + u_simu;               // overloaded X.rplus(u) = X * exp(u)
+
+        /// then we receive noisy gps measurement - - - - - - - - - - - - - - - -
+        y_noise = y_sigmas * Array2d::Random();             // simulate measurement noise
+
+        y = X_simulation.translation();                     // position measurement, before adding noise
+        y = y + y_noise;                                    // position measurement, noisy
+
+
+
+
+        //// II. Estimation ###############################################################################
+
+        /// First we move - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+        X = X.plus(u_est, J_x, J_u);                        // X * exp(u), with Jacobians
+
+        P = J_x * P * J_x.transpose() + J_u * U * J_u.transpose();
+
+
+        /// Then we correct using the gps position - - - - - - - - - - - - - - -
+
+        // transform covariance from right to left invariant
+        AdX = X.adj();
+        AdXinv = X.inverse().adj();
+        P_L = AdX * P * AdX.transpose();                    // from eq. (54) in the paper
```

Comment:
What is the purpose of this? some questions arise:
- is it clear that you can swap invariances from left to right and back? Does the filter keep the nice properties of invariant filters when you do this?
- For what I know, you usually get left-invariance, or right-invariance, and not both. Is this operation allowing you to have both? or are you "cheating" in some way?
- this is a costly operation in KF. Is it worth it?
- When are you converting the covariance back to right invariant? If you keep converting only to left, then at each loop you convert again.

Overall, this needs some explanations.

Reply:
Please see the comment below.